compiler/ETA/CodeGen/Constr.hs

```diff
@@ -41,8 +41,8 @@ cgTopRhsCon dflags id dataCon args = (cgIdInfo, genCode)
         -- typeFt = obj (tyConClass dflags (dataConTyCon dataCon))
         genCode = do
           loads <- mapM getArgLoadCode . getNonVoids $ zip maybeFields args
-          defineField $ mkFieldDef [Private, Static] qClName closureType
-          let field = mkFieldRef modClass qClName closureType
+          defineField $ mkFieldDef [Public, Static] qClName dataFt
```

Comment:
We typically avoid typing fields or method arguments with anything other than `closureType` since we want to be able to rewrite closures at runtime. The JIT will take care of remembering the types for us so I don't think it'll be a huge difference. Note that I did do this actively in the first iteration of the codegen and I had to revert to using `closureType` because there are corner cases where making it too typed is a problem. 

Let me make that concrete with an example. Suppose we have

```
data Test = Test 1 2 3
test :: Test
test = Test 1 2 3
```

And suppose we later redefine it to an `Int` (in cases like an interpreter)

```
test :: Int
test = 1
```

If it was an `StgClosure`, this is very easy to change since it's just a matter of resetting the field. If it was more specific, it would be impossible to place it in the field since Java doesn't allow unsafe coercing.